subsys/bluetooth/mesh/adv.c

```diff
@@ -93,35 +95,126 @@ static struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *buf_poo
 	adv->type         = type;
 	adv->tag          = tag;
 	adv->xmit         = xmit;
+	adv->prio         = prio;
 
 	return buf;
 }
 
-struct net_buf *bt_mesh_adv_create(enum bt_mesh_adv_type type,
-				   enum bt_mesh_adv_tag tag,
-				   uint8_t xmit, k_timeout_t timeout)
+struct net_buf *bt_mesh_adv_relay_create(uint8_t prio, uint8_t xmit)
 {
 #if defined(CONFIG_BT_MESH_RELAY)
-	if (tag & BT_MESH_RELAY_ADV) {
-		return bt_mesh_adv_create_from_pool(&relay_buf_pool,
-						    adv_relay_pool, type,
-						    tag, xmit, timeout);
+	sys_sfnode_t *prev, *curr, *lowest_prev = NULL, *lowest = NULL;
+	uint8_t prio_cur = prio;
+	struct net_buf *buf;
+
+	buf = bt_mesh_adv_create_from_pool(&relay_buf_pool, adv_relay_pool,
+					   BT_MESH_ADV_DATA, BT_MESH_RELAY_ADV,
+					   xmit, prio, K_NO_WAIT);
+	if (buf) {
+		return buf;
 	}
-#endif
 
-	return bt_mesh_adv_create_from_pool(&adv_buf_pool, adv_local_pool, type,
-					    tag, xmit, timeout);
+	if (!IS_ENABLED(CONFIG_BT_MESH_RELAY_PRIORITY) || !prio_cur) {
+		return NULL;
+	}
+
+	/* Try to release the buffer from the low priority queue to the high priority message.
+	 * For the same low priority message, the one that takes the longest buffer time will
+	 * be released first.
+	 */
+	k_mutex_lock(&bt_mesh_relay_mutex, K_FOREVER);
+
+	SYS_SFLIST_FOR_EACH_NODE(&bt_mesh_relay_queue.data_q, curr) {
```

Comment:
I guess it is not good to access internal entries of `k_queue`. Perhaps there could be something like `K_QUEUE_FOR_EACH`?

Reply:
Your suggestion is meaningful.

Reply:
I think this needs to be addressed.